reStream.sh

```diff
@@ -86,6 +79,76 @@ if ! ssh_cmd true; then
     exit 1
 fi
 
+rm_version="$(ssh_cmd cat /sys/devices/soc0/machine)"
+
+# We cannot read the framebuffer directory in the reMarkable 2, but we know the
+# address of the pointer to the framebuffer for some reMarkable 2 versions.
+# This will need to be changed each time xochitl updates.
+rm2_getpointer() {
+    xochitl_version="$(ssh_cmd "sed -n 's/^REMARKABLE_RELEASE_VERSION=//p' /usr/share/remarkable/update.conf")"
+    case "$xochitl_version" in
+        "2.3."*)
+            echo "4121024"
+            ;;
+        *) # use last known version if we cannot find a match
+            echo "WARNING: unknown reMarkable 2 release version" >&2
+            echo "consider updating if reStream doesn't work" >&2
+            echo "4121024"
+            ;;
+    esac
+}
+
+# Print the instruction for reading a memory location.
+# Using dd with bs=1 is too slow, so we first carve out the pages our desired
+# bytes are located in, and then we trim the resulting data with what we need.
+memread_instr() {
+    page_size=4096
+    ifile="$1"  # input (device) file
+    offset="$2" # offset in bytes to start reading
+    count="$3"  # byte count to read
+
+    window_start_blocks="$((offset / page_size))"
+    window_skip_bytes="$((offset % page_size))"
+    window_length_blocks="$((count / page_size + 1))"
+    "dd if=$ifile bs=$page_size skip=$window_start_blocks count=$window_length_blocks 2>/dev/null | dd if=/dev/stdin skip=$window_skip_bytes bs=$count count=1 2>/dev/null"
+}
+
+case "$rm_version" in
+    "reMarkable 1.0")
+        width=1408
+        height=1872
+        bytes_per_pixel=2
+        pixel_format="rgb565le"
+        # calculate how much bytes the window is
+        window_bytes="$((width * height * bytes_per_pixel))"
+        # read the first $window_bytes of the framebuffer
+        head_fb0="dd if=/dev/fb0 count=1 bs=$window_bytes 2>/dev/null"
+        ;;
+    "reMarkable 2.0")
+        width=1872
+        height=1404
+        bytes_per_pixel=1
+        pixel_format="gray8"
+        pid="$(ssh_cmd pidof xochitl)"
+        pointer="$(rm2_getpointer)"
+        read_address="addr=\$($(memread_instr "/proc/$pid/mem" "$pointer" 4) | hexdump | awk '{print \$3\$2}') && printf '%d' \$((16#\$addr))"
```

Comment:
I am not a bash expert, but I suspect we may want to add "echo" before the second dollar sign?